python/TestHarness/schedulers/QueueManager.py

```diff
@@ -0,0 +1,403 @@
+            try:
+                os.makedirs(self.getWorkingDir(job_container))
+            except OSError, ex:
+                if ex.errno == errno.EEXIST: pass
+                else: raise
+
+        # Move into the newly created test directory
+        current_pwd = os.getcwd()
+        os.chdir(self.getWorkingDir(job_container))
+
+        # Copy files (unless they are listed in "no_copy")
+        for file in os.listdir('../'):
+            if os.path.isfile('../' + file) and file != self.options.input_file_name and \
+               (not tester.specs.isValid('no_copy') or file not in tester.specs['no_copy']) and \
+               (not 'no_copy_pattern' in self.params.keys() or pattern.match(file) is None) and \
+               not os.path.exists(file) and \
+               os.path.splitext(file)[1] != '':
+                shutil.copy('../' + file, '.')
+
+        # Copy directories
+        if 'copy_files' in template.keys():
+            for file in template['copy_files']:
+                if os.path.isfile('../' + file):
+                    if not os.path.exists(file):
+                        shutil.copy('../' + file, '.')
+                elif os.path.isdir('../' + file):
+                    try:
+                        shutil.copytree('../' + file, file)
+                    except OSError, ex:
+                        if ex.errno == errno.EEXIST: pass
+                        else: raise Exception()
+
+        # Create symlinks (do last, to allow hard copy above to be default)
+        if tester.specs.isValid('link_files'):
+            for file in tester.specs['link_files']:
+                if os.path.exists('../' + file):
+                    try:
+                        os.symlink('../' + file, file)
+                    except OSError, ex:
+                        if ex.errno == errno.EEXIST: pass
+                        else: raise Exception()
+
+        # return to the previous directory
+        os.chdir(current_pwd)
+
+    def downstreamFinished(self, job_container):
+        """ Return bool for given job_container for all downstream jobs finished """
+        job_dag = job_container.getOriginalDAG()
+        downstream_jobs = job_dag.all_downstreams(job_container)
+        for job in downstream_jobs:
+            test_unique = self.getUnique(job)
+            json_job = self.getData(test_unique, isFinished=True)
+
+            # None would mean the upstream job never launched
+            if 'isFinished' not in json_job.keys() or json_job['isFinished'] is False:
+                return False
+        return True
+
+    def queueJobs(self, status_jobs=[], run_jobs=[]):
+        """
+        Entry point into the QueueManager. Supply queueJobs with a list
+        of jobs to either launch, or print statuses for.
+
+        Syntax:
+
+           Launch job:
+            queueJobs(run_jobs=[list of jobs])
+
+           Print status:
+            queueJobs(status_jobs=[list of jobs])
```

Comment:
Honestly, these really should just be separate methods. Since `queueJobs` is your entry point, this method would just turn around and call these two methods as appropriate. No need to combine these functions into a single method with a giant "if" statement.